### PR TITLE
Support digressions when steps without url conditions are used for escalate/end

### PR DESCRIPTION
**Pull request by peterszerzo**

### CODE REVIEW

File: packages/journey-manager/src/index.ts [hunk 1]
```diff
@@ -211,17 +211,29 @@ export const run = (props: RunProps): RunOutput => {
 
   const triggeredSteps = getTriggeredSteps(props.config.conversationId);
 
-  // If all triggers have a URL constraint set and none match the current URL, fire a digression calback
+  /**
+   * Digression detection
+   * - if all triggers have a URL constraint set and none match the current URL, fire a digression callback
+   * - notable exception: if the step is used as escalation or end as well, missing URL constraint is ok.
```
Comment (michaelglass) [inline packages/journey-manager/src/index.ts:217]: rewriting to confirm I understand what you're saying
```suggestion
   * - notable exception: ignore URL constraints if if the step is used as an escalation or end.
```

File: packages/journey-manager/src/index.ts [hunk 2]
```diff
@@ -211,17 +211,29 @@ export const run = (props: RunProps): RunOutput => {
...
+  const isDigressionDetectable = !Object.entries(props.triggers).some(
+    ([stepId, trigger]) => {
+      return (
+        trigger.urlCondition == null &&
+        !(stepId === props.ui?.escalationStep || stepId === props.ui?.endStep)
+      );
+    },
+  );
+
+  if (isDigressionDetectable && !urlConditions.some(matchesUrlCondition)) {
```
Comment (michaelglass) [inline packages/journey-manager/src/index.ts:236]: 💭 minor nit: you can avoid all of the computation by checking if `onDigression` is set before checking for a digression.

File: packages/journey-manager/src/index.ts [hunk 3]
```diff
@@ -211,17 +211,29 @@ export const run = (props: RunProps): RunOutput => {
...
     (trigger) => trigger.urlCondition,
   );
 
-  if (
-    Object.keys(props.triggers).length === urlConditions.length &&
-    !urlConditions.some(matchesUrlCondition)
-  ) {
+  // If there are any steps for which there is no URL condition while also not being used for escalation or end,
+  // the package assumes that a digression cannot be reliably detected.
+  const isDigressionDetectable = !Object.entries(props.triggers).some(
```
Comment (michaelglass) [inline packages/journey-manager/src/index.ts:227]: suggest rewriting this as an `every` for readability

```suggestion
  const isDigressionDetectable = !Object.entries(props.triggers).every(
    ([stepId, trigger]) => {
      return (
        // every step has a URL condition
        trigger.urlCondition != null ||
        // unless it's an escalation step or end step
        stepId === props.ui?.escalationStep ||
        stepId === props.ui?.endStep
      );
    },
  );
```

**Review by michaelglass — DISMISSED** [review]
some stylistic suggestions. Take them or leave them.

File: packages/journey-manager/src/index.ts (hunk 2 above)
Comment (peterszerzo) [reply packages/journey-manager/src/index.ts:236]: In the future there will be automatic digression handling (later today actually), I think I would leave this as it is.

File: packages/journey-manager/src/index.ts (hunk 3 above)
Comment (peterszerzo) [reply packages/journey-manager/src/index.ts:227]: @michaelglass you meant to take out the `!` from the `Object.entries` here right?

Comment (michaelglass) [reply packages/journey-manager/src/index.ts:227]: whooops, yeah!

**Review by jakub-nlx — DISMISSED** [review]

File: packages/journey-manager/src/index.ts (hunk 3 above)
Comment (peterszerzo) [reply packages/journey-manager/src/index.ts]: I agree, so nice @michaelglass 😍

**Review by jakub-nlx — APPROVED** [review]

**Pull request merged**